Assets/_Scripts/CurveController.cs

```diff
@@ -1,96 +1,138 @@
-            UpdateALLMarkersRotation();
-        }
+        //TODO What was this meant to be anyway?
+        GameEventSystem.Instance.Raise<UpdateSplineEvent, bool>(new UpdateSplineEvent(markerList));
     }
 
-    public void AddMarker()
+    private void AddMarkerHandler(onRightClickPerformed onRightClickPerformed)
     {
-        Vector2 markerPosition = Vector2.zero; //Initializing default markerPosition
+        Vector2 markerPosition = onRightClickPerformed.mousePosition;
+        List<int> affectedMarkersRotationIndex = new List<int>();
 
-        if (splineController.curveHovered) //Checks if will add a marker inbetween.
+        if (
+            onRightClickPerformed.mouseHitAllArray3D != null
+            && onRightClickPerformed.mouseHitAllArray3D.Length != 0
+        ) //Checks if will add a marker inbetween.
         {
-            AddMarker_INBETWEEN();
+            //Raising insert Marker event to get info to where the marker needs to be on the spline.
+            MarkerDataStruct markerData = GameEventSystem.Instance.Raise<
+                InsertMarkerEvent,
+                MarkerDataStruct
+            >(new InsertMarkerEvent(onRightClickPerformed.mousePosition));
+            InsertMarker(markerData);
+
+            affectedMarkersRotationIndex.Add(markerData.markerIndex);
+            affectedMarkersRotationIndex.Add(markerData.markerIndex - 1);
+            affectedMarkersRotationIndex.Add(markerData.markerIndex + 1);
+
+            //AddMarkerEvent with markerData index and position.
+            GameEventSystem.Instance.Raise<AddMarkerEvent>(
+                new AddMarkerEvent(markerData.markerPosition, markerData.markerIndex)
+            );
         }
         else //If not, then add to the end.
         {
-            AddMarker_TAIL(markerPosition);
+            AddMarker(markerPosition);
+
+            //AddMarkerEvent with last index and markerPosition.
+            int lastIndex = markerList.Count - 1;
+            affectedMarkersRotationIndex.Add(lastIndex);
+            if (lastIndex > 0)
+            {
+                affectedMarkersRotationIndex.Add(lastIndex - 1);
+            }
+            GameEventSystem.Instance.Raise<AddMarkerEvent>(
+                new AddMarkerEvent(markerPosition, lastIndex)
+            );
         }
+
+        UpdateMarkersRotation(affectedMarkersRotationIndex);
     }
 
-    public void AddMarker_TAIL(Vector2 markerPosition)
+    private void AddMarker(Vector2 markerPosition)
     {
-        DeselectAllMarkers();
-        GameObject newMarker;
-        // Vector2 markerPosition;
-
-        if (markerPosition == Vector2.zero) //This basically doesn't allow the user to put markers on (0,0) which he should be allowed. It's a dirty FIX.
-        {
-            markerPosition = inputManager.GetWorldMouseLocation2D();
-        }
-
-        newMarker = Instantiate(marker, markerPosition, Quaternion.identity, this.transform);
+        GameObject newMarker = Instantiate(
```

Comment:
As I mentioned, we should always work with MarkerEntity instead of the GameObject. Therefore, I'd suggest instantiating the MarkerEntity rather than GameObject.